sktime/classification/early_classification/_teaser.py

```diff
@@ -0,0 +1,423 @@
+                    decisions = [True for _ in range(n_instances)]
+                elif self._svms[i] is not None:
+                    decisions = self._svms[i].predict(X_svm[i]) == 1
+                else:
+                    decisions = [False for _ in range(n_instances)]
+
+                # record consecutive class decisions
+                state_info = [
+                    (
+                        i,
+                        state_info[n][1] + 1
+                        if decisions[n] and train_preds[i][n] == state_info[i][2]
+                        else 1
+                        if decisions[n]
+                        else 0,
+                        train_preds[i][n],
+                    )
+                    if not finished[n]
+                    else state_info[n]
+                    for n in range(n_instances)
+                ]
+
+                # safety decisions
+                finished = [
+                    True if state_info[n][1] >= g else False for n in range(n_instances)
+                ]
+
+            accuracy = (
+                np.sum(
+                    [
+                        1 if state_info[i][2] == self._class_dictionary[y[i]] else 0
+                        for i in range(n_instances)
+                    ]
+                )
+                / n_instances
+            )
+            earliness = (
+                1
+                - np.sum(
+                    [
+                        self.classification_points[state_info[i][0]] / series_length
+                        for i in range(n_instances)
+                    ]
+                )
+                / n_instances
+            )
+            hm = (2 * accuracy * earliness) / (accuracy + earliness)
+
+            if hm > best_hm:
+                best_hm = hm
+                self._consecutive_predictions = g
+
+        return self
+
+    def _predict(self, X):
+        rng = check_random_state(self.random_state)
+        return np.array(
+            [
+                self.classes_[int(rng.choice(np.flatnonzero(prob == prob.max())))]
+                for prob in self._predict_proba(X)
+            ]
+        )
+
+    def _predict_proba(self, X):
+        _, _, series_length = X.shape
+        idx = self._classification_point_dictionary.get(series_length, -1)
+        if idx == -1:
+            raise ValueError(
+                f"Input series length does not match the classification points produced"
+                f" in fit. Current classification points: {self._classification_points}"
+            )
+
+        m = getattr(self._estimators[idx], "predict_proba", None)
+        if callable(m):
+            return self._estimators[idx].predict_proba(X)
+        else:
+            dists = np.zeros((X.shape[0], self.n_classes_))
+            preds = self._estimators[idx].predict(X)
+            for i in range(0, X.shape[0]):
+                dists[i, self._class_dictionary[preds[i]]] = 1
+            return dists
```

Comment:
Not sure, if I understand correctly. Isnt a dist the inverse of a probability? i.e. higher is worse.

Maybe consider using: CalibratedClassifierCV() ?

Reply:
Holdover from tsml I have ingrained, I think. Short for probability distributions. Will rename them to probas.

We didn't have much success with probability calibration when we looked into it with HIVE-COTE. Classifiers which don't produce probability estimates such as ROCKET did not perform better with any of the techniques we tried. 

That sounds like a wrapper for an estimator? The user can pass it as the base estimator for TEASER if required. If you think it improves WEASEL/MUSE then we can use it as a default.

Reply:
Yes, CalibratedClassifierCV  comes from: https://scikit-learn.org/stable/modules/generated/sklearn.calibration.CalibratedClassifierCV.html
It will add the `predict_proba` to a classifier.

So, I am not sure, what should predict_proba() return when the estiamtor has no predict_proba available?

Reply:
Given we have to train a classifier for each selected time stamp, I think adding a cross-validation for each on top of that is a bit much. Did the Java version return valid probabilities?

In tsml (and currently here) we just return 1 for the predicted class as a default method for predict_proba, i.e. [0, 1, 0, 0] if the 2nd class is predicted in predict.

Reply:
Agreed! 

Hmm, TEASER in Java is based on WEASEL, which returns probabilities. 

I am not pushing this, but maybe it would be nice to separate this:

1. If the classifier has a predict_proba(), return the probabilities
2. If the classifier has only a decision_function(), return distances

I think, sklearn handles it like that: https://github.com/scikit-learn/scikit-learn/blob/7e1e6d09b/sklearn/svm/_base.py#L807

But, if it is done like that in sktime so far, I am not pushing it :)

Reply:
I'm unsure if any of our classifiers have a decision_function() implementation actually, not entirely sure what it does.

It is done similarly to this currently, not saying thats the best way to go about it however. WEASEL has predict_proba so not a big issue by default, I think if you change it to an estimator without predict_proba that's the user's issue 🙂.

Reply:
Rocket with RidgeClassifierCV, for example, has only the decision function... :) 

decision_function() predicts confidence scores  in sklearn - i.e. it simply returns the distance. 

I agree that we do not have to over-design the issue. Is it possible to check, if a classifier passed to TEASER has the predict_proba method as a prerequisite?

Reply:
So, eventually check if 
m = getattr(self._estimators[idx], "predict_proba", None)
is true and given an error if not.

Reply:
Probably a safer option that just returning bad probabilities, I agree. Will include it in the next push.